locations/jclouds/src/main/java/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -1305,118 +1326,148 @@ public UserCreation(LoginCredentials creds, List<Statement> statements) {
         String user = getUser(config);
         String explicitLoginUser = config.get(LOGIN_USER);
         String loginUser = groovyTruth(explicitLoginUser) ? explicitLoginUser : (image != null && image.getDefaultCredentials() != null) ? image.getDefaultCredentials().identity : null;
         Boolean dontCreateUser = config.get(DONT_CREATE_USER);
         Boolean grantUserSudo = config.get(GRANT_USER_SUDO);
-        String publicKeyData = LocationConfigUtils.getPublicKeyData(config);
-        String privateKeyData = LocationConfigUtils.getPrivateKeyData(config);
-        String explicitPassword = config.get(PASSWORD);
-        String password = groovyTruth(explicitPassword) ? explicitPassword : Identifiers.makeRandomId(12);
+        OsCredential credential = LocationConfigUtils.getOsCredential(config);
+        credential.checkNoErrors().logAnyWarnings();
+        String passwordToSet = Strings.isNonBlank(credential.getPassword()) ? credential.getPassword() : Identifiers.makeRandomId(12);
         List<Statement> statements = Lists.newArrayList();
         
         if (groovyTruth(dontCreateUser)) {
-            // TODO For dontCreateUser, we probably only want to treat it special if user was explicitly supplied
-            // (rather than it just being the default config key value). If user was explicit, then should
-            // set the password + authorize the key for that user. Presumably the caller knows that this
-            // user pre-exists on the given VM image.
+            // dontCreateUser:
+            // if caller has not specified a user, we'll just continue to use the loginUser;
+            // if caller *has*, we set up our credentials assuming that user and credentials already exist
+            
             if (!groovyTruth(user)) {
-                // loginCreds result will be null; use creds returned by jclouds on the node
+                // createdUserCreds returned from this method will be null; 
+                // we will use the creds returned by jclouds on the node
                 LOG.info("Not setting up any user (subsequently using loginUser {})", user, loginUser);
                 config.put(USER, loginUser);
                 
             } else {
-                LOG.info("Not creating user {}, and not setting its password or authorizing keys", user);
-                
-                if (privateKeyData != null) {
-                    loginCreds = LoginCredentials.builder().user(user).privateKey(privateKeyData).build();
-                } else if (explicitPassword != null) {
-                    loginCreds = LoginCredentials.builder().user(user).password(password).build();
+                LOG.info("Not creating user {}, and not installing its password or authorizing keys (assuming it exists)", user);
+
+                if (credential.isUsingPassword()) {
+                    createdUserCreds = LoginCredentials.builder().user(user).password(credential.get()).build();
```

Comment:
I'd prefer `credential.getPassword()` rather than `credential.get()` - it's much more explicit, rather than relying on `.get()` returning password because `isUsingPassword()`. Same for line below.